src/spi.cpp

```diff
@@ -0,0 +1,254 @@
+      break;
+    }
+    case 3: {
+      m_peripheral_register = get_spi_reg(peripheral::spi3);
+      m_peripheral_id = peripheral::spi3;
+      pin clock(peripheral::gpio_b, 3);
+      pin data_in(peripheral::gpio_b, 5);
+      pin data_out(peripheral::gpio_b, 4);
+
+      clock.function(pin::pin_function::alternate7)
+        .open_drain(false)
+        .resistor(pin_resistor::none);
+      data_in.function(pin::pin_function::alternate7)
+        .open_drain(false)
+        .resistor(pin_resistor::none);
+      data_out.function(pin::pin_function::alternate7)
+        .open_drain(false)
+        .resistor(pin_resistor::none);
+      break;
+    }
+    case 4: {
+      m_peripheral_register = get_spi_reg(peripheral::spi4);
+      m_peripheral_id = peripheral::spi4;
+      pin clock(peripheral::gpio_b, 0);
+      pin data_in(peripheral::gpio_b, 8);
+      pin data_out(peripheral::gpio_a, 12);
+
+      clock.function(pin::pin_function::alternate6)
+        .open_drain(false)
+        .resistor(pin_resistor::none);
+      data_in.function(pin::pin_function::alternate6)
+        .open_drain(false)
+        .resistor(pin_resistor::none);
+      data_out.function(pin::pin_function::alternate6)
+        .open_drain(false)
+        .resistor(pin_resistor::none);
+      break;
+    }
+    case 5: {
+      m_peripheral_register = get_spi_reg(peripheral::spi5);
+      m_peripheral_id = peripheral::spi5;
+      pin clock(peripheral::gpio_b, 3);
+      pin data_in(peripheral::gpio_b, 5);
+      pin data_out(peripheral::gpio_b, 4);
+
+      clock.function(pin::pin::pin_function::alternate6)
+        .open_drain(false)
+        .resistor(pin_resistor::none);
+      data_in.function(pin::pin::pin_function::alternate6)
+        .open_drain(false)
+        .resistor(pin_resistor::none);
+      data_out.function(pin::pin_function::alternate6)
+        .open_drain(false)
+        .resistor(pin_resistor::none);
+      break;
+    }
+    default:
+      // "Supported spi busses are 1-5!";
+      hal::safe_throw(hal::operation_not_supported(this));
+  }
+  power(m_peripheral_id).on();
+  spi::driver_configure(p_settings);
+}  // namespace hal::lpc40
+
+spi::~spi()
+{
+  power(m_peripheral_id).off();
+}
+
+void spi::driver_configure(const settings& p_settings)
+{
+  auto* reg = reinterpret_cast<spi_reg_t*>(m_peripheral_register);
+
+  // Set SPI to master mode by clearing
+  bit_modify(reg->cr1).set(control_register1::master_selection);
+
+  // Setup operating frequency
+  // TODO replace input clock with a get_frequency instruction (#16)
+  using namespace hal::literals;
+  const auto input_clock = 16.0_MHz;
+  const auto clock_divider = input_clock / p_settings.clock_rate;
+  auto prescaler = static_cast<std::uint16_t>(clock_divider);
+  if (prescaler <= 1) {
+    prescaler = 2;
+  } else if (prescaler > 256) {
+    hal::safe_throw(hal::operation_not_supported(this));
+  }
+
+  [[maybe_unused]] uint16_t baud_control = 15 - std::countl_zero(prescaler);
```

Comment:
```suggestion
uint16_t baud_control = 15 - std::countl_zero(prescaler);
```

It looks like this is used so we can eliminate the maybe unused section.